src/transformers/rc4.ts

```diff
@@ -0,0 +1,101 @@
+/// <reference path='../../../third_party/sha1/sha1.d.ts' />
+/// <reference path='../../../third_party/simple-rc4/simple-rc4.d.ts' />
+/// <reference path='../../../third_party/typings/browser.d.ts' />
+
+import logging = require('../logging/logging');
+import rc4 = require('simple-rc4');
+import sha1 = require('crypto/sha1');
+import transformer = require('./transformer');
+
+const log = new logging.Log('rc4 transformer');
+
+// Accepted in serialised form by configure().
+export interface Config {
+  key: string
+}
+
+const KEY_LENGTH_BYTES = 16;
+const R_LENGTH_BYTES = 8;
+const TRUNCATED_IV_LENGTH_BYTES = 8;
+
+// Creates a sample (non-random) config, suitable for testing.
+export function sampleConfig(): Config {
+  return {
+    key: new Buffer(KEY_LENGTH_BYTES).fill(0).toString('hex')
+  };
+}
+
+// Fast, uniformly random transformer with minimal length expansion, via RC4.
+// From a proposal by kpdyer.
+//
+// Terminology:
+//  - K: 128-bit session key, negotiated through the signalling channel
+//  - RANDOM(N): a cryptographically secure N-byte string
+//  - X[n,...,m]: the (m - n + 1) bytes starting at byte-index n of X
+//  - |X|: length, in bytes, of X
+//
+// Pseudo-code:
+//   def Encrypt(K, P):
+//     R = RANDOM(8)
+//     IV = SHA1(K || R)
+//     RET R || RC4(IV[0,...,7], P)
+//
+//   def Decrypt(K1, C):
+//     R = C[0,...,7]
+//     IV = SHA1(K1 || R)
+//     RET RC4(IV[0,...,7], C[8,...,|C | -1])
+export class Rc4Transformer implements transformer.Transformer {
+  private key_: Buffer;
+
+  public constructor() {
+    this.configure(JSON.stringify(sampleConfig()));
+  }
+
+  public configure = (json: string): void => {
+    try {
+      const config = <Config>JSON.parse(json);
+      if (config.key === undefined) {
+        throw new Error("must set key parameter");
+      }
+      const key = new Buffer(config.key, 'hex');
+      if (key.byteLength !== KEY_LENGTH_BYTES) {
+        throw new Error('keys must be ' + KEY_LENGTH_BYTES + ' bytes in length');
+      }
+      this.key_ = key;
+    } catch (e) {
+      throw new Error('could not parse config: ' + e.message);
+    }
+  }
+
+  // Applies RC4(IV[0,...,7] to bytes, as described in the pseudocode above.
+  private update_ = (r: Buffer, bytes:Buffer): void => {
+    // TODO: use the crypto module and avoid string conversion
+    const iv = sha1.str_sha1(Buffer.concat([this.key_, r]).toString('binary'));
+    const truncatedIv = new Buffer(iv, 'binary').slice(0, TRUNCATED_IV_LENGTH_BYTES);
+    new rc4(truncatedIv).update(bytes);
+  }
+
+  public transform = (ab: ArrayBuffer): ArrayBuffer[] => {
+    const p = new Buffer(ab);
+
+    const r = new Buffer(R_LENGTH_BYTES);
+    crypto.getRandomValues(r);
+
+    this.update_(r, p);
+
+    return [Buffer.concat([r, p]).buffer];
+  }
+
+  public restore = (ab: ArrayBuffer): ArrayBuffer[] => {
```

Comment:
Yeah, this is an interesting line of inquiry ... but I'm very happy to have the incremental improvement.  In the future, it'll be interesting to hear from researchers who are considering active probing attacks.